controllers/persistentvolumeclaim_controller.go

```diff
@@ -37,83 +36,18 @@ func (r *PersistentVolumeClaimReconciler) Reconcile(ctx context.Context, req ctr
 		return ctrl.Result{}, err
 	}
 
-	if pvc.DeletionTimestamp == nil {
-		return ctrl.Result{}, nil
-	}
-
-	needFinalize := false
-	onlyTopoLVM := true
-	// Due to bug #310, we need to de-dup TopoLVM finalizers.
-	for _, f := range pvc.Finalizers {
-		if f == topolvm.PVCFinalizer {
-			needFinalize = true
-			continue
-		}
-		onlyTopoLVM = false
-	}
-	if !needFinalize {
-		return ctrl.Result{}, nil
-	}
-
-	// Requeue until other finalizers complete their jobs.
-	if !onlyTopoLVM {
-		return ctrl.Result{
-			Requeue:      true,
-			RequeueAfter: 10 * time.Second,
-		}, nil
-	}
-
-	pvc.Finalizers = nil
-	if err := r.Update(ctx, pvc); err != nil {
-		log.Error(err, "failed to remove finalizer", "name", pvc.Name)
-		return ctrl.Result{}, err
-	}
-
-	// sleep shortly to wait StatefulSet controller notices PVC deletion
-	time.Sleep(100 * time.Millisecond)
-
-	pods, err := r.getPodsByPVC(ctx, pvc)
-	if err != nil {
-		log.Error(err, "unable to fetch PodList for a PVC", "pvc", pvc.Name, "namespace", pvc.Namespace)
-		return ctrl.Result{}, err
-	}
-	for _, pod := range pods {
-		err := r.Delete(ctx, &pod, client.GracePeriodSeconds(1))
-		if err != nil {
-			log.Error(err, "unable to delete Pod", "name", pod.Name, "namespace", pod.Namespace)
+	var removed bool
+	pvc.Finalizers, removed = removeTopoLVMFinalizer(pvc)
+	if removed {
+		if err := r.Update(ctx, pvc); err != nil {
```

Comment:
Why not do these things in `removeTopoLVMFinalizer()`?

Reply:
Thanks. I moved these lines to `removeTopoLVMFinalizer`.